doc/dev/dialects.rst

```diff
@@ -0,0 +1,422 @@
+that other Catalyst components can depend on:
+
+.. code-block:: cmake
+
+    add_mlir_library(MLIROpenQASM
+        OpenQASM.cpp
+
+        DEPENDS
+        MLIROpenQASMIncGen
+    )
+
+Note the naming scheme: ``MLIROpenQASM`` is a name of our choice for the dialect build target, while
+``MLIROpenQASMIncGen`` is a target automatically provided by the `add_mlir_dialect` function from the
+provided TableGen file name (``OpenQASM``). The latter represents the generation of C++ files from
+TableGen.
+
+.. warning::
+
+    For any newly added ``CMakeLists.txt``, be sure to add it to its parent CMake file with
+    ``add_subdirectory(<name of new folder>)``. In this case, both ``mlir/include/CMakeLists.txt``
+    and ``mlir/lib/CMakeLists.txt`` will need to be updated with ``add_subdirectory(OpenQASM)``.
+
+
+Using the dialect
+=================
+
+MLIR's standard tool for testing dialects and compiler passes is the ``opt`` tool (inherited from
+LLVM). The tool parses a program in the textual MLIR format, applies arbitrary passes, and prints
+the transformed program back out. Parsing and printing out a program without any transformations
+is also referred to as "round-tripping". Let's see if we can pass this first test with our dialect!
+
+Catalyst comes with its own version of the opt tool, ``quantum-opt``, preloaded with all builtin
+MLIR dialects and transformations, as well all additional compiler components developed for
+Catalyst specifically. Find the file located at ``mlir/tools/quantum-opt/quantum-opt.cpp`` and add
+the following two lines to it:
+
+.. code-block:: c++
+
+    // ...
+    #include "OpenQASM/OpenQASM.h"  // add me
+
+    int main(int argc, char **argv)
+    {
+        // ...
+        registry.insert<catalyst::openqasm::OpenQASMDialect>();  // add me
+
+        // ...
+    }
+
+Similarly, update the corresponding ``mlir/tools/quantum-opt/CMakeLists.txt`` to include the build
+target for our dialect as a dependent library:
+
+.. code-block:: cmake
+
+    # ...
+    set(LIBS
+        # ...
+        MLIROpenQASM  # add me
+    )
+
+    # ...
+
+That's it! We can now build our additions with the rest of the dialects and test them out. Assuming
+Catalyst has already been built successfully at least once, simply run:
+
+.. code-block:: console
+
+    make dialects
+
+Save the following test file somewhere and run it through the ``quantum-opt`` tool:
+
+.. code-block:: mlir
+
+    func.func @my_circuit(%q0 : !oq.qubit) {
+        %phi = arith.constant 0.3 : f64
+
+        oq.RZ(%phi) %q0 : !oq.qubit
+
+        func.return
+    }
+
+.. code-block:: console
+
+    ./mlir/build/bin/quantum-opt my_test_file.mlir
+
+You should see the same code in the input file printed back out to you:
+
+.. code-block::
+
+    func.func @my_circuit(%q0 : !oq.qubit) {
+        %phi = arith.constant 0.3 : f64
+
+        oq.RZ(%phi) %q0 : !oq.qubit
+
+        func.return
+    }
+
```

Comment:
```suggestion
module {
  func.func @my_circuit(%arg0: !oq.qubit) {
    %cst = arith.constant 3.000000e-01 : f64
    oq.RZ(%cst) %arg0 : !oq.qubit
    return
  }
}
```

Reply:
That was the actual output in my machine, variable names were changed, as well as the float value notation, among other things. No big deal but might be worth mentioning that refactoring.

Reply:
True, there can be small changes (like the implicit module op being printed, and SSA names which are not part of the IR and thus are not preserved, line breaks which are also not part of the IR, etc). I'll change it to the version you saw :)